In [1]:
#Importing the libraries
import numpy as np
import pandas as pd

In [ ]:
#Reading the Goodreads data into gd
gd = pd.read_csv('goodreads_books.csv')
gd.head()

In [4]:
gd.columns

Index(['id', 'title', 'link', 'series', 'cover_link', 'author', 'author_link',
       'rating_count', 'review_count', 'average_rating', 'five_star_ratings',
       'four_star_ratings', 'three_star_ratings', 'two_star_ratings',
       'one_star_ratings', 'number_of_pages', 'date_published', 'publisher',
       'original_title', 'genre_and_votes', 'isbn', 'isbn13', 'asin',
       'settings', 'characters', 'awards', 'amazon_redirect_link',
       'worldcat_redirect_link', 'recommended_books', 'books_in_series',
       'description'],
      dtype='object')

In [7]:
gd.drop(['cover_link', 'number_of_pages', 'amazon_redirect_link', 'worldcat_redirect_link'], axis=1, inplace=True)
gd

id                                              title  \
0        630104                                       Inner Circle   
1          9487                                  A Time to Embrace   
2       6050894                                           Take Two   
3         39030                                          Reliquary   
4           998  The Millionaire Next Door: The Surprising Secr...   
...         ...                                                ...   
52194  15743072                             The Stranger I Married   
52195  18143905     The Opposite of Loneliness: Essays and Stories   
52196   1466878                                  Sadako will leben   
52197    630103                                        Confessions   
52198   6922622                      Going Rogue: An American Life   

                                                    link               series  \
0      https://www.goodreads.com//book/show/630104.In...         (Private #5)   
1      https://www.goodreads.com//book/show/9487.A_Ti...   (Timeless Love #2)   
2      https://www.goodreads.com//book/show/6050894-t...  (Above the Line #2)   
3      https://www.goodreads.com//book/show/39030.Rel...      (Pendergast #2)   
4      https://www.goodreads.com//book/show/998.The_M...                  NaN   
...                                                  ...                  ...   
52194  https://www.goodreads.com//book/show/15743072-...                  NaN   
52195  https://www.goodreads.com//book/show/18143905-...                  NaN   
52196  https://www.goodreads.com//book/show/1466878.S...                  NaN   
52197  https://www.goodreads.com//book/show/630103.Co...         (Private #4)   
52198  https://www.goodreads.com//book/show/6922622-g...                  NaN   

                                          author  \
0                      Kate Brian, Julian Peploe   
1                                Karen Kingsbury   
2                                Karen Kingsbury   
3                 Douglas Preston, Lincoln Child   
4            Thomas J. Stanley, William D. Danko   
...                                          ...   
52194                                 Sylvia Day   
52195                              Marina Keegan   
52196                              Karl Bruckner   
52197                                 Kate Brian   
52198  Sarah Palin, Lynn Vincent, Dewey Whetsell   

                                             author_link  rating_count  \
0      https://www.goodreads.com/author/show/94091.Ka...          7597   
1      https://www.goodreads.com/author/show/3159984....          4179   
2      https://www.goodreads.com/author/show/3159984....          6288   
3      https://www.goodreads.com/author/show/12577.Do...         38382   
4      https://www.goodreads.com/author/show/659.Thom...         72168   
...                                                  ...           ...   
52194  https://www.goodreads.com/author/show/19823.Sy...         13295   
52195  https://www.goodreads.com/author/show/6153502....         42609   
52196  https://www.goodreads.com/author/show/687674.K...          1583   
52197  https://www.goodreads.com/author/show/94091.Ka...          8815   
52198  https://www.goodreads.com/author/show/2040815....         11359   

       review_count  average_rating  five_star_ratings  ...  \
0               196            4.03               3045  ...   
1               177            4.35               2255  ...   
2               218            4.23               3000  ...   
3              1424            4.01              12711  ...   
4              3217            4.04              27594  ...   
...             ...             ...                ...  ...   
52194           805            3.87               4220  ...   
52195          4282            3.84              12461  ...   
52196            64            4.17                673  ...   
52197           258            4.09               3760 

# Find Duplicates in titles

In [11]:
gd['title'].duplicated().sum()

2514

We have a lot of duplicates in the title column. Have to explore more what caused this duplication.

In [12]:
dup_mask = gd['title'].duplicated()
gd.loc[dup_mask, 'title']

723        Collected Poems
1162       Come Back to Me
1256     Limits of Destiny
1305               Twisted
1383                Unsaid
               ...        
52181          Untouchable
52185               Chosen
52186                Shift
52193          Blue Heaven
52197          Confessions
Name: title, Length: 2514, dtype: object

In [31]:
gd.loc[gd['title'] == 'Collected Poems', :]

id            title  \
334    133380  Collected Poems   
723    149505  Collected Poems   
5134   111879  Collected Poems   
30872  317024  Collected Poems   
32329  321538  Collected Poems   
35990   22990  Collected Poems   
39371  125209  Collected Poems   
51396  367802  Collected Poems   

                                                    link series  \
334    https://www.goodreads.com//book/show/133380.Co...    NaN   
723    https://www.goodreads.com//book/show/149505.Co...    NaN   
5134   https://www.goodreads.com//book/show/111879.Co...    NaN   
30872  https://www.goodreads.com//book/show/317024.Co...    NaN   
32329  https://www.goodreads.com//book/show/321538.Co...    NaN   
35990  https://www.goodreads.com//book/show/22990.Col...    NaN   
39371  https://www.goodreads.com//book/show/125209.Co...    NaN   
51396  https://www.goodreads.com//book/show/367802.Co...    NaN   

                                                  author  \
334                Edna St. Vincent Millay, Norma Millay   
723                              Ted Hughes, Paul Keegan   
5134                      Philip Larkin, Anthony Thwaite   
30872                                   Patrick Kavanagh   
32329                       W.H. Auden, Edward Mendelson   
35990  Federico García Lorca, Christopher Maurer, Ca...   
39371        Robert Lowell, David Gewanter, Frank Bidart   
51396                                       Dylan Thomas   

                                             author_link  rating_count  \
334    https://www.goodreads.com/author/show/33998.Ed...          8094   
723    https://www.goodreads.com/author/show/996.Ted_...          1236   
5134   https://www.goodreads.com/author/show/64716.Ph...          7959   
30872  https://www.goodreads.com/author/show/30098.Pa...           300   
32329  https://www.goodreads.com/author/show/233593.W...          5918   
35990  https://www.goodreads.com/author/show/44150.Fe...          3098   
39371  https://www.goodreads.com/author/show/36912.Ro...          2658   
51396  https://www.goodreads.com/author/show/57189.Dy...         15377   

       review_count  average_rating  five_star_ratings  ...  \
334             140            4.24               3897  ...   
723              36            4.09                529  ...   
5134            229            4.22               3676  ...   
30872            16            4.21                134  ...   
32329            84            4.28               2909  ...   
35990            73            4.47               1883  ...   
39371            34            4.07               1086  ...   
51396           264            4.20               6924  ...   

                                         genre_and_votes        isbn  \
334       Poetry 887, Classics 52, Fiction 11, Womens 11  0060908890   
723    Poetry 219, Literature 4, European Literature-...  0374125384   
5134   Poetry 1007, Classics 24, European Literature-...  0374529205   
30872  Poetry 74, Cultural-Ireland 8, European Litera...  0393006948   
32329  Poetry 688, Classics 27, Literature 14, Europe...  0679731970   
35990  Poetry 387, European Literature-Spanish Litera...  0374126151   
39371  Poetry 260, Literature-American 5, Fiction 3, ...  0374530327   
51396  Poetry 1379, Classics 73, Fiction 36, Literatu...  0811202054   

              isbn13  asin settings    characters  \
334    9780060908898   NaN      NaN           NaN   
723    9780374125387   NaN      NaN           NaN   
5134   9780374529208   NaN      NaN           NaN   
30872  9780393006940   NaN      NaN           NaN   
32329  9780679731979   NaN      NaN           NaN   
35990  9780374126155   NaN      NaN           NaN   
39371  9780374530327   NaN      NaN           NaN   
51396  9780811202053   NaN      NaN  Dylan Thomas   

                                        awards  \
334                                        NaN   
723                                        NaN   
5134                                       NaN 

For 'Collected Poems', the title is used for different books with the same generic name to name a book of collection of poems. So we have to alter it to include the authors' names so as to differentiate it. 

In [32]:
gd.loc[gd['title'] == 'Limits of Destiny', :]

id              title  \
1098  26212359  Limits of Destiny   
1256  28480416  Limits of Destiny   
1429  26008216  Limits of Destiny   
1745  26128174  Limits of Destiny   
2026  27223433  Limits of Destiny   

                                                   link  \
1098  https://www.goodreads.com//book/show/26212359-...   
1256  https://www.goodreads.com//book/show/28480416-...   
1429  https://www.goodreads.com//book/show/26008216-...   
1745  https://www.goodreads.com//book/show/26128174-...   
2026  https://www.goodreads.com//book/show/27223433-...   

                      series              author  \
1098  (Limits of Destiny #1)  Sharlyn G. Branson   
1256  (Limits of Destiny #5)  Sharlyn G. Branson   
1429  (Limits of Destiny #2)  Sharlyn G. Branson   
1745  (Limits of Destiny #3)  Sharlyn G. Branson   
2026  (Limits of Destiny #4)  Sharlyn G. Branson   

                                            author_link  rating_count  \
1098  https://www.goodreads.com/author/show/14161119...           493   
1256  https://www.goodreads.com/author/show/14161119...           114   
1429  https://www.goodreads.com/author/show/14161119...           183   
1745  https://www.goodreads.com/author/show/14161119...           155   
2026  https://www.goodreads.com/author/show/14161119...           149   

      review_count  average_rating  five_star_ratings  ...  \
1098           147            3.86                183  ...   
1256            57            4.50                 78  ...   
1429            79            4.32                106  ...   
1745            68            4.38                 93  ...   
2026            56            4.47                 98  ...   

                                        genre_and_votes  isbn  isbn13  \
1098  Romance 28, Adult Fiction-Erotica 9, Contempor...   NaN     NaN   
1256                            Adult Fiction-Erotica 3   NaN     NaN   
1429  Romance-Contemporary Romance 4, Romance 3, Adu...   NaN     NaN   
1745          Romance-Contemporary Romance 5, Romance 2   NaN     NaN   
2026          Romance 4, Romance-Contemporary Romance 3   NaN     NaN   

            asin     settings characters awards  \
1098         NaN  Switzerland        NaN    NaN   
1256  B01A4MDTOW          NaN        NaN    NaN   
1429  B012ELKRF0          NaN        NaN    NaN   
1745  B0143DRNL8          NaN        NaN    NaN   
2026  B016QTOFEA          NaN        NaN    NaN   

                                      recommended_books  \
1098  50150140, 50150140, 43688586, 53876583, 500072...   
1256  52120990, 52120990, 40976731, 38371364, 523053...   
1429  52120990, 52120990, 45014184, 48712798, 289294...   
1745  52120990, 52120990, 44151683, 45014184, 487127...   
2026  52120990, 52120990, 38371364, 40302987, 523053...   

                             books_in_series  \
1098  26008216, 26128174, 27223433, 28480416   
1256  26212359, 26008216, 26128174, 27223433   
1429  26212359, 26128174, 27223433, 28480416   
1745  26212359, 26008216, 27223433, 28480416   
2026  26212359, 26008216, 26128174, 28480416   

                                            description  
1098  Alexia Welson is very beautiful and men do not...  
1256  Does love have the power to overcome all obsta...  
1429  The detonation of a bomb in Kraftberg’s hote...  
1745  Back in Switzerland, Alexia is forced to face ...  
2026  What are the thresholds of pain tolerance?  Al...  

[5 rows x 27 columns]

Some books have the same name as they belong to the same book series.

In [33]:
gd.loc[gd['title'] == 'Twisted', :]

id    title                                               link  \
458    16029550  Twisted  https://www.goodreads.com//book/show/16029550-...   
1305     123106  Twisted  https://www.goodreads.com//book/show/123106.Tw...   
2445    9893331  Twisted  https://www.goodreads.com//book/show/9893331-t...   
22314  17185680  Twisted  https://www.goodreads.com//book/show/17185680-...   
24880  14760308  Twisted  https://www.goodreads.com//book/show/14760308-...   
25548  17969365  Twisted  https://www.goodreads.com//book/show/17969365-...   
31151  12362225  Twisted  https://www.goodreads.com//book/show/12362225-...   
32510  23598959  Twisted  https://www.goodreads.com//book/show/23598959-...   
37163  35341537  Twisted  https://www.goodreads.com//book/show/35341537-...   
47213  21446984  Twisted  https://www.goodreads.com//book/show/21446984-...   
52019  17977710  Twisted  https://www.goodreads.com//book/show/17977710-...   

                              series                 author  \
458         (Twisted Fairy Tales #1)        Jenny  Phillips   
1305                             NaN  Laurie Halse Anderson   
2445        (Pretty Little Liars #9)           Sara Shepard   
22314         (Dark Protectors #5.5)        Rebecca Zanetti   
24880                            NaN             Amity Hope   
25548                      (Torn #2)          K.A. Robinson   
31151                   (Tracers #5)          Laura Griffin   
32510      (New York Crime Kings #2)             Skyla Madi   
37163  (Imperfect Love Kindle World)             Mandi Beck   
47213                   (Twisted #1)          Lola Smirnova   
52019                   (Tangled #2)             Emma Chase   

                                             author_link  rating_count  \
458    https://www.goodreads.com/author/show/5784961....            55   
1305   https://www.goodreads.com/author/show/10003.La...         33883   
2445   https://www.goodreads.com/author/show/93970.Sa...         27902   
22314  https://www.goodreads.com/author/show/4149342....          4103   
24880  https://www.goodreads.com/author/show/6428483....          2468   
25548  https://www.goodreads.com/author/show/6880914....         10229   
31151  https://www.goodreads.com/author/show/84181.La...          3006   
32510  https://www.goodreads.com/author/show/6554179....           653   
37163  https://www.goodreads.com/author/show/9500789....           990   
47213  https://www.goodreads.com/author/show/7857144....           183   
52019  https://www.goodreads.com/author/show/6551005....         29829   

       review_count  average_rating  five_star_ratings  ...  \
458              25            4.29                 32  ...   
1305           3093            3.83               9786  ...   
2445           1267            4.03              11536  ...   
22314           271            4.20               1738  ...   
24880           207            3.97                766  ...   
25548           584            4.04               3613  ...   
31151           209            4.17               1169  ...   
32510            72            4.16                262  ...   
37163           250            4.34                519  ...   
47213            86            3.85                 65  ...   
52019          2755            4.02              10661  ...   

                                         genre_and_votes        isbn  \
458    Fairy Tales-Fairy Tale Retellings 3, Fantasy-P...         NaN   
1305   Young Adult 1004, Realistic Fiction 289, Ficti...  0670061018   
2445                        Young Adult 469, Mystery 259  0062081012   
22314  Fantasy-Paranormal 146, Romance-Paranormal Rom...         NaN   
24880  Young Adult 57, Romance 47, Contemporary 20, R...         NaN   
25548  New Adult 132, Romance 102, Academic-College 3...  1476752176   
31151  Romance-Romantic Suspense 108, Romance 51, Sus...         NaN   
32510  Warfare-Fighters 14, Dark 6, New Adult 5, Adul...         NaN   
37163  Novell

For 'Twisted' we see that different books can have the same names belonging to different series or standalone books.

So in conclusion, we based on our exploration, we conclude that duplication of titles is okay as it is common.

# ISBN duplicates

In [21]:
gd['isbn'].duplicated().sum()

11882

In [22]:
isbn_dup_mask = gd['isbn'].duplicated()
gd[isbn_dup_mask]

id                                              title  \
19     16116760                                            Silence   
22     18523711                                           Darkness   
23     17234659                       Ten Thousand Skies Above You   
24     25906415                 Liam: Midsummer's Magic Bonus Book   
56     25975141  Dragon of Legend: Destiny: Fantasy Dragon Adve...   
...         ...                                                ...   
52140  26017870                                              Kaden   
52141  40948486                                             George   
52158  16068341                              The Truth About Alice   
52173  17700010                                    Falling into Us   
52180  35965782                                 The Lost Continent   

                                                    link               series  \
19     https://www.goodreads.com//book/show/16116760-...         (Silence #1)   
22     https://www.goodreads.com//book/show/18523711-...    (New Species #12)   
23     https://www.goodreads.com//book/show/17234659-...        (Firebird #2)   
24     https://www.goodreads.com//book/show/25906415-...                  NaN   
56     https://www.goodreads.com//book/show/25975141-...                  NaN   
...                                                  ...                  ...   
52140  https://www.goodreads.com//book/show/26017870-...                  NaN   
52141  https://www.goodreads.com//book/show/40948486-...                  NaN   
52158  https://www.goodreads.com//book/show/16068341-...                  NaN   
52173  https://www.goodreads.com//book/show/17700010-...         (Falling #2)   
52180  https://www.goodreads.com//book/show/35965782-...  (Wings of Fire #11)   

                  author                                        author_link  \
19       Natasha Preston  https://www.goodreads.com/author/show/6548349....   
22        Laurann Dohner  https://www.goodreads.com/author/show/3190846....   
23          Claudia Gray  https://www.goodreads.com/author/show/1192311....   
24       Emmie Lou Kates  https://www.goodreads.com/author/show/7142452....   
56        Angelika Meyer  https://www.goodreads.com/author/show/7399064....   
...                  ...                                                ...   
52140     Synne Jakobsen  https://www.goodreads.com/author/show/14136590...   
52141          Alex Gino  https://www.goodreads.com/author/show/12881739...   
52158   Jennifer Mathieu  https://www.goodreads.com/author/show/6549106....   
52173     Jasinda Wilder  https://www.goodreads.com/author/show/6521173....   
52180  Tui T. Sutherland  https://www.goodreads.com/author/show/374504.T...   

       rating_count  review_count  average_rating  five_star_ratings  ...  \
19            28383          2300            3.96              11872  ...   
22            10256           587            4.28               5062  ...   
23            17770          2638            4.08               6234  ...   
24               17             0            5.00                 17  ...   
56              126            19            4.72                101  ...   
...             ...           ...             ...                ...  ...   
52140            76            13            4.28                 47  ...   
52141         27951          5582            4.06              10174  ...   
52158         13463          2487            3.65               2716  ...   
52173         17865          1294            4.24               8357  ...   
52180          5172           365            4.62               3694  ...   

                                         genre_and_votes  isbn  isbn13  \
19     Romance 184, Young Adult 172, Contemporary 92,...   NaN     NaN   
22     Fantasy-Paranormal 249, Romance 198, Romance-P...   NaN     NaN   
23     Young Adult 617, Science Fiction 414, Romance ...   NaN     NaN   
24                           

In [23]:
gd['isbn'].isnull().sum()

11883

We see that all these duplicates are actually null values. Now let us check if all the books have atleast one of the isbn formats.

In [25]:
isbn_mask = gd['isbn'].isnull() & gd['isbn13'].isnull()
isbn_mask.sum()

11883

In [28]:
len(gd['id'].unique())

52199

All those 11883 books dont have either isbn or isbn_13, but we have unique ids for all those books.